src/Components/Components/src/SupplyParameterFromPersistentComponentStateProviderServiceCollectionExtensions.cs

```diff
@@ -0,0 +1,53 @@
+
+namespace Microsoft.AspNetCore.Components;
+
+/// <summary>
+/// Enables component parameters to be supplied from <see cref="PersistentComponentState"/> with <see cref="SupplyParameterFromPersistentComponentStateAttribute"/>.
+/// </summary>
+public static class SupplyParameterFromPersistentComponentStateProviderServiceCollectionExtensions
+{
+    /// <summary>
+    /// Enables component parameters to be supplied from <see cref="PersistentComponentState"/> with <see cref="SupplyParameterFromPersistentComponentStateAttribute"/>..
+    /// </summary>
+    /// <param name="services">The <see cref="IServiceCollection"/>.</param>
+    /// <returns>The <see cref="IServiceCollection"/>.</returns>
+    public static IServiceCollection AddSupplyValueFromPersistentComponentStateProvider(this IServiceCollection services)
+    {
+        services.TryAddEnumerable(ServiceDescriptor.Scoped<ICascadingValueSupplier, SupplyParameterFromPersistentComponentStateValueProvider>());
+        return services;
+    }
+
+    /// <summary>
+    /// Saves <typeparamref name="TService"/> state when the application is persisting state and restores it at the appropriate time automatically.
+    /// </summary>
+    /// <remarks>
+    /// Only public properties annotated with <see cref="SupplyParameterFromPersistentComponentStateAttribute"/> are persisted and restored.
+    /// </remarks>
+    /// <typeparam name="TService">The service type to register for persistence.</typeparam>
+    /// <param name="services">The <see cref="IServiceCollection"/>.</param>
+    /// <param name="componentRenderMode">The <see cref="IComponentRenderMode"/> to register the service for.</param>
+    /// <returns>The <see cref="IServiceCollection"/>.</returns>
+    public static IServiceCollection AddPersistentService<[DynamicallyAccessedMembers(JsonSerialized)] TService>(
+        this IServiceCollection services,
+        IComponentRenderMode componentRenderMode)
+    {
+        // This method does something very similar to what we do when we register root components, except in this case we are registering services.
+        // We collect a list of all the registrations on during static rendering mode and push those registrations to interactive mode.
+        // When the interactive mode starts, we retrieve the registry, and this triggers the process for restoring all the states for all the services.
+        // The process for retrieving the services is the same as we do for root components.
+        // We look for the assembly in the current list of loaded assemblies.
+        // We look for the type inside the assembly.
+        // We resolve the service from the DI container.
+        // We loop through the properties in the type and try to restore the properties that have SupplyParameterFromPersistentComponentState on them.
+        services.TryAddEnumerable(ServiceDescriptor.Singleton<IPersistentComponentRegistration>(new PersistentComponentRegistration<TService>(componentRenderMode)));
```

Comment:
Am I right in thinking this *only* registers the `IPersistentComponentRegistration`, and doesn't register `TService` itself in DI? If so that might be confusing as you'd have code like this:

```cs
builder.Services.AddPersistentService<MyService>();
```

People would reasonably think that registers `MyService` in DI and hence can `@inject MyService ...` elsewhere. If that doesn't work, it would suggest the API shape might be a bit weird.

Arguably this should be something configured on `RazorComponentsServiceOptions` instead of `IServiceCollection`.

Reply:
I think we can hang this off of `IRazorComponentBuilder` and call it `RegisterPersistentService`.

Reply:
Ah, we can't do that for the same reason we have all the `public static IServiceCollection AddSupplyValueFromQueryProvider(this IServiceCollection services)` and so on methods. This lives in `Microsoft.AspNetCore.Components`.

The best we can do is to move it somewhere pubternal, not make it an extension method and then have an extension method that calls the "pubternal" API hanging from IRazorComponentBuilder